File: middle of the linked list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* middleNode(ListNode* head) {
        vector<ListNode*> links;
        int counter = 0;
        while(head != NULL)
        {
            links.push_back(head);
            head = head->next;
            counter++;
        }
        counter = (counter/2 );
        return links[counter];
    }
};

File: MiddleOfTheLinkedList.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:
        
        # size = 0
        # temp = head

        # while temp != None:
        #     temp = temp.next
        #     size += 1

        # temp = head

        # for i in range(size//2):
        #     temp = temp.next

        # return temp

        # temp = head
        # nums = []

        # while temp != None:
        #     nums.append(temp)
        #     temp = temp.next

        # return nums[len(nums)//2]

        slow = head
        fast = head

        while fast != None:
            fast = fast.next
            if fast == None:
                break
            fast = fast.next
            slow = slow.next

        return slow

     
